internal/core/component_factory.rs

```diff
@@ -0,0 +1,259 @@
+// SPDX-License-Identifier: GPL-3.0-only OR LicenseRef-Slint-commercial
+
+#![warn(missing_docs)]
+
+//! This module contains the `ComponentFactory` and a `Component`. Both are
+//! exposed to the C API.
+
+use core::ffi::c_void;
+use core::fmt::Debug;
+
+use alloc::{boxed::Box, rc::Rc};
+
+use crate::{
+    api::ComponentHandle,
+    component::{ComponentRc, ComponentVTable},
+};
+
+/// A `Component` built by a `ComponentFactory`
+#[repr(C)]
+pub struct Component(ComponentRc);
+
+impl<T: ComponentHandle> From<T> for Component
+where
+    T::Inner: vtable::HasStaticVTable<ComponentVTable> + 'static,
+{
+    fn from(value: T) -> Self {
+        Self(vtable::VRc::into_dyn(value.as_weak().inner().upgrade().unwrap()))
+    }
+}
+
+impl Component {
+    pub(crate) fn component_rc(self: Self) -> ComponentRc {
+        self.0
+    }
+
+    pub(crate) fn clone_internal(&self) -> Component {
+        Component(self.0.clone())
+    }
+}
+
+pub(crate) mod component_factory_data {
+    use super::{c_void, Component};
+
+    /// A internal helper needed to implement `ComponentFactory` in different languages.
+    #[derive(Clone, PartialEq)]
+    pub(crate) struct ComponentFactoryData {
+        pub build: unsafe extern "C" fn(*mut c_void) -> *mut Component,
+        pub data: *mut c_void,
+        pub drop: unsafe extern "C" fn(*mut c_void),
+    }
+
+    #[allow(unsafe_code)]
+    impl ComponentFactoryData {
+        pub fn build(&self) -> Option<Component> {
+            let result = unsafe { (self.build)(self.data) };
+            if result.is_null() {
+                None
+            } else {
+                let r = unsafe { (*result).clone_internal() };
+                Some(r)
+            }
+        }
+    }
+
+    impl Default for ComponentFactoryData {
+        fn default() -> Self {
+            let data = Box::new(RustData { factory: Box::new(|| None) });
+            Self {
+                build: slint_component_factory_builder_call,
+                data: Box::into_raw(data) as *mut c_void,
+                drop: slint_component_factory_builder_drop,
+            }
+        }
+    }
+
+    #[allow(unsafe_code)]
+    impl Drop for ComponentFactoryData {
+        fn drop(&mut self) {
+            unsafe { (self.drop)(self.data) };
+            self.data = core::ptr::null_mut();
+        }
+    }
+
+    pub(crate) struct RustData {
+        pub(crate) factory: Box<dyn Fn() -> Option<Component>>,
+    }
+
+    #[allow(unsafe_code)]
+    pub(crate) unsafe extern "C" fn slint_component_factory_builder_drop(data: *mut c_void) {
+        assert!(!data.is_null());
+        drop(Box::from_raw(data as *mut RustData));
+    }
+
+    #[allow(unsafe_code)]
+    pub(crate) unsafe extern "C" fn slint_component_factory_builder_call(
+        data: *mut c_void,
+    ) -> *mut Component {
+        let data = (data as *mut RustData).as_mut().unwrap();
+        if let Some(result) = (data.factory)() {
+            Box::into_raw(Box::new(result))
```

Comment:
Afaics this box is leaked.

Why the indirection through a raw pointer and a box instead of returning a vrc dyn here (by means of an out pointer)?

Reply:
We do the out parameter pattern with an Rc here: https://github.com/slint-ui/slint/blob/ea8b5d31176bbaecd37a4eca3b8c11a93bb7ac56/api/cpp/lib.rs#L23

The option can be done by letting the pointer remain null.

that way you avoid returning raw pointers with unclear ownership.